receipe-app-api/src/recipe/recipe.service.ts

```diff
@@ -1,71 +1,78 @@
-import { Injectable } from '@nestjs/common';
+import { Inject, Injectable } from '@nestjs/common';
+import { InjectModel } from '@nestjs/mongoose';
+import { Model } from 'mongoose';
+import { isMongoId } from 'class-validator';
 
-import { DetailRecipeModel } from './models';
+import { ProductsService } from '../products/products.service';
+import { CategoriesService } from '../categories/categories.service';
 
-const ImageMock = 'https://picsum.photos/500/500';
+import { Recipe } from './schemas';
+import { CreateRecipeDto } from './dto';
 
 @Injectable()
 export class RecipeService {
-  async getRecipeById(id: string): Promise<DetailRecipeModel> {
-    await new Promise(resolve => setTimeout(resolve, 1500));
-
-    /* eslint-disable */
-    return {
-      id: id,
-      title: 'Product 1 ',
-      image: ImageMock,
-      rating: 4,
-      kCal: 500,
-      weight: 1300,
-      time: '1:50',
-      description:
-        'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Dictumst vestibulum rhoncus est pellentesque elit ullamcorper dignissim cras. Ac turpis egestas sed tempus.',
+  @InjectModel(Recipe.name) private recipeModel: Model<Recipe>;
+  @Inject(CategoriesService) private readonly categoriesService: CategoriesService;
+  @Inject(ProductsService) private readonly productsService: ProductsService;
+
+  findOneByTitle(title: string): Promise<Recipe | null> {
+    return this.recipeModel.findOne({ title }).exec();
+  }
+
+  findOneById(id: string): Promise<Recipe | null> {
+    if (!isMongoId(id)) {
+      return Promise.resolve(null);
+    }
+
+    return this.recipeModel.findOne({ _id: id }).exec();
+  }
+
+  async create(createRecipeDto: CreateRecipeDto): Promise<Recipe> {
+    const found = await this.findOneByTitle(createRecipeDto.title);
+
+    if (found) {
+      throw new Error(`Recipe ${createRecipeDto.title} already exists`);
+    }
+
+    for (const category of createRecipeDto.categories) {
+      if (!await this.categoriesService.findOneByTitle(category)) {
+        throw new Error(`Category ${category} does not exist`);
+      }
+    }
+
+    let kCal = 0;
+    let fats = 0;
+    let carbs = 0;
+    let proteins = 0;
+    for (const ingredient of createRecipeDto.ingredients) {
+      const product = await this.productsService.findOneByTitle(ingredient.title);
+
+      if (!product) {
```

Comment:
why do we need to check the products list, if we select products from the products list, when creating a recipe?

Reply:
in order to ensure that the given product exists. In case if given product is not found, the app is not able to calculate nutritions for the whole recipe